File: internal/catalog/catalog.go
package catalog

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"log"
	"net/http"
	"os"
	"path/filepath"
	"regexp"
	"strconv"
	"time"

	"df-hud/internal/dfclient"
)

// The public game data. get_allstats.php?printvars=1 is unauthenticated, about
// 1MB, and changes only when the game updates, so it is fetched at most once a
// day and cached on disk. Same feed df-allstats-watcher already polls.
//
// One thing in it matters to df-hud:
//
//   - exp_lvl<N>: the XP that must accumulate WITHIN level N-1 to reach level
//     N. Not an assumption - the game's own checkIfLevelUp (base.js:448-478)
//     compares df_exp against exp_lvl[level+1] and carries the remainder over
//     as the new df_exp on level-up. That is exactly what makes a cumulative
//     XP counter reconstructable, and therefore what makes XP/hr immune to the
//     level-up reset in df_exp.
//
// The feed also carries two map grids - zones_<x>_<y>_name, 13x13 neighbourhood
// names, and areas_<x>_<y>_*, a 39x39 grid of streets and buildings - and both
// used to be parsed and cached here against the day the transform from
// df_positionx/y onto them was solved. They are gone, because they are not the
// city df-hud draws: the city is 59x55 with 1716 blocks in it, 39x39 is 1521
// cells, and no landmark exists on both sides to pin them together. Nothing in
// the saved web client references either key. See knowledge/city-map.md for
// where the map actually comes from.
const catalogSchemaVersion = 1

const (
	// catalogMaxBody caps the download. The real body is ~1MB; 8MB is room for
	// growth while still refusing to buffer something pathological.
	catalogMaxBody = 8 << 20
	// catalogMinBody rejects a truncated response before it becomes a half
	// catalog with a plausible-looking XP table.
	catalogMinBody = 64 << 10
	catalogMax     = 10000 // sanity bound on level numbers
)

type Catalog struct {
	SchemaVersion int       `json:"schema_version"`
	FetchedAt     time.Time `json:"fetched_at"`

	// MaxLevel is the highest level the feed has a threshold for (415 today,
	// which matches the hard cap in checkIfLevelUp).
	MaxLevel int `json:"max_level"`
	// ExpToReach is indexed by level: ExpToReach[n] is the feed's exp_lvl<n>,
	// the XP needed within level n-1 to reach level n. Entries 0 and 1 are
	// unused, so the index is the level and no off-by-one arithmetic is needed
	// at the call sites.
	ExpToReach []int64 `json:"exp_to_reach"`

	// cumExp[n] is the total XP earned since level 1 by the moment you reach
	// level n, i.e. the prefix sum of ExpToReach. Derived, not persisted:
	// storing it would double the file to save a loop over 415 integers.
	cumExp []int64
}

var expKeyRe = regexp.MustCompile(`^exp_lvl(\d+)$`)

// parseCatalog builds a catalog from an already-parsed feed. It takes the map
// rather than the raw body so the same code serves the network path, the disk
// path and the test fixture.
func parseCatalog(vars map[string]string, fetchedAt time.Time) (*Catalog, error) {
	c := &Catalog{SchemaVersion: catalogSchemaVersion, FetchedAt: fetchedAt}

	// --- XP table ---
	exp := map[int]int64{}
	for key, raw := range vars {
		m := expKeyRe.FindStringSubmatch(key)
		if m == nil {
			continue
		}
		level, err := strconv.Atoi(m[1])
		if err != nil || level < 1 || level > catalogMax {
			continue
		}
		v, err := strconv.ParseInt(raw, 10, 64)
		if err != nil {
			return nil, fmt.Errorf("catalog: %s = %q is not a number", key, raw)
		}
		if v < 0 {
			return nil, fmt.Errorf("catalog: %s = %d is negative", key, v)
		}
		if level > c.MaxLevel {
			c.MaxLevel = level
		}
		exp[level] = v
	}
	if c.MaxLevel < 2 {
		return nil, errors.New("catalog: the feed carries no exp_lvl table")
	}
	// A hole in the table would silently corrupt every cumulative sum past it,
	// which would show up as an XP rate that is quietly wrong rather than
	// obviously broken. Refuse instead.
	c.ExpToReach = make([]int64, c.MaxLevel+1)
	for level := 2; level <= c.MaxLevel; level++ {
		v, ok := exp[level]
		if !ok {
			return nil, fmt.Errorf("catalog: exp_lvl%d is missing, so the XP table has a hole "+
				"and cumulative totals past level %d would be wrong", level, level-1)
		}
		c.ExpToReach[level] = v
	}

	c.finish()
	return c, nil
}

// finish computes the derived prefix sums. Must be called after parsing and
// after loading from disk, since cumExp is not persisted.
func (c *Catalog) finish() {
	c.cumExp = make([]int64, len(c.ExpToReach))
	var total int64
	for level := 2; level < len(c.ExpToReach); level++ {
		total += c.ExpToReach[level]
		c.cumExp[level] = total
	}
}

// ExpNeeded is the threshold shown as "df_exp / needed" in the game's own
// sidebar (base.js:1660): the XP required to leave the given level.
func (c *Catalog) ExpNeeded(level int) (int64, bool) {
	if level < 1 || level+1 >= len(c.ExpToReach) {
		return 0, false // level 415 is the cap and has nothing above it
	}
	return c.ExpToReach[level+1], true
}

// CumulativeXP is total XP earned since level 1, the counter XP/hr differences.
// It exists because df_exp resets to the carry-over on every level-up
// (checkIfLevelUp writes currentExp - neededExp), so a naive delta goes
// negative at each boundary. This sum is continuous across level-ups by
// construction: the XP that vanishes from df_exp reappears in the prefix sum.
func (c *Catalog) CumulativeXP(level int, expInLevel int64) (int64, bool) {
	if level < 1 || level >= len(c.cumExp) {
		return 0, false
	}
	return c.cumExp[level] + expInLevel, true
}

// Summary is one log line at startup, and the fastest way to notice that a
// game update changed the feed's shape.
func (c *Catalog) Summary() string {
	return fmt.Sprintf("levels 2..%d", c.MaxLevel)
}

// fetchCatalog downloads and parses the feed. GET, no credentials: this
// endpoint is public, which is the whole reason Block Info needs no login.
func fetchCatalog(ctx context.Context, hc *http.Client, feedURL, userAgent string) (*Catalog, error) {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, feedURL, nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("User-Agent", userAgent)

	resp, err := hc.Do(req)
	if err != nil {
		return nil, fmt.Errorf("catalog: %w", err)
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("catalog: HTTP %s", resp.Status)
	}
	raw, err := io.ReadAll(io.LimitReader(resp.Body, catalogMaxBody))
	if err != nil {
		return nil, fmt.Errorf("catalog: reading body: %w", err)
	}
	body := string(raw)
	if dfclient.LooksLikeHTML(body) {
		return nil, errors.New("catalog: got an HTML page instead of the feed (Cloudflare or an outage)")
	}
	if len(raw) < catalogMinBody {
		return nil, fmt.Errorf("catalog: body is only %d bytes, far short of the ~1MB feed; "+
			"treating it as truncated rather than parsing a partial table", len(raw))
	}
	vars, err := dfclient.ParseFlash(body)
	if err != nil {
		return nil, fmt.Errorf("catalog: %w", err)
	}
	return parseCatalog(vars, time.Now())
}

// loadCatalogFile reads the disk cache. A missing file returns (nil, nil): that
// is a first run, not a failure. A corrupt or old-schema file is quarantined
// and also returns (nil, nil), so df-hud re-fetches instead of crash-looping -
// the same discipline as state.json.
func loadCatalogFile(path string) (*Catalog, error) {
	data, err := os.ReadFile(path)
	if errors.Is(err, os.ErrNotExist) {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	var c Catalog
	if err := json.Unmarshal(data, &c); err != nil || c.SchemaVersion != catalogSchemaVersion || len(c.ExpToReach) < 3 {
		aside := fmt.Sprintf("%s.corrupt-%d", path, time.Now().Unix())
		if renameErr := os.Rename(path, aside); renameErr != nil {
			return nil, fmt.Errorf("catalog cache unusable and could not be moved aside: %w", renameErr)
		}
		log.Printf("catalog: cache unusable, moved to %s, will re-fetch", aside)
		return nil, nil
	}
	c.finish()
	return &c, nil
}

func saveCatalogFile(path string, c *Catalog) error {
	data, err := json.Marshal(c)
	if err != nil {
		return err
	}
	if err := os.MkdirAll(filepath.Dir(path), 0o700); err != nil {
		return err
	}
	tmp := path + ".tmp"
	f, err := os.Create(tmp)
	if err != nil {
		return err
	}
	if _, err := f.Write(data); err == nil {
		err = f.Sync()
	}
	if closeErr := f.Close(); err == nil {
		err = closeErr
	}
	if err != nil {
		os.Remove(tmp)
		return err
	}
	return os.Rename(tmp, path)
}

// ensureCatalog returns a usable catalog, in order of preference:
//
//  1. a cached copy younger than maxAge - no network at all
//  2. a fresh download, written to the cache
//  3. the stale cache, if the download failed
//
// Step 3 is the point. This data changes only when the game updates, so a
// week-old XP table is still correct, and an outage or a Cloudflare
// interstitial should not cost you the HUD.
func ensureCatalog(ctx context.Context, hc *http.Client, path, feedURL, userAgent string, maxAge time.Duration) (*Catalog, error) {
	cached, err := loadCatalogFile(path)
	if err != nil {
		log.Printf("catalog: %v", err) // not fatal; fall through to the network
	}
	if cached != nil && time.Since(cached.FetchedAt) < maxAge {
		return cached, nil
	}

	fresh, err := fetchCatalog(ctx, hc, feedURL, userAgent)
	if err != nil {
		if cached != nil {
			log.Printf("catalog: refresh failed (%v); using the cached copy from %s",
				err, cached.FetchedAt.Format(time.RFC3339))
			return cached, nil
		}
		return nil, err
	}
	if err := saveCatalogFile(path, fresh); err != nil {
		log.Printf("catalog: could not write the cache: %v", err) // usable anyway
	}
	return fresh, nil
}

// Parse builds a catalog from an allstats variable map.
func Parse(vars map[string]string, fetchedAt time.Time) (*Catalog, error) {
	return parseCatalog(vars, fetchedAt)
}

// Ensure returns a fresh-enough catalog, with stale-cache fallback.
func Ensure(ctx context.Context, hc *http.Client, path, feedURL, userAgent string, maxAge time.Duration) (*Catalog, error) {
	return ensureCatalog(ctx, hc, path, feedURL, userAgent, maxAge)
}

// parseFlash remains package-local for the focused tests and delegates to the
// protocol package that owns the wire format.
func parseFlash(body string) (map[string]string, error) { return dfclient.ParseFlash(body) }

File: src/catalog.rs
//! XP table from the public allstats feed. Keys are `exp_lvlN`.

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::io::Read;
use std::path::Path;
use std::time::Duration;

use crate::dfclient;

const SCHEMA_VERSION: i32 = 1;
const MAX_BODY: u64 = 8 << 20;
const MIN_BODY: usize = 64 << 10;
const LEVEL_CAP: i32 = 10000;

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct Catalog {
    pub schema_version: i32,
    pub fetched_at: DateTime<Utc>,
    pub max_level: i32,
    pub exp_to_reach: Vec<i64>,
    #[serde(skip)]
    cum_exp: Vec<i64>,
}

impl Catalog {
    fn finish(&mut self) {
        self.cum_exp = vec![0; self.exp_to_reach.len()];
        let mut total = 0i64;
        for level in 2..self.exp_to_reach.len() {
            total += self.exp_to_reach[level];
            self.cum_exp[level] = total;
        }
    }

    pub fn exp_needed(&self, level: i32) -> Option<i64> {
        let i = (level as usize).checked_add(1)?;
        if level < 1 || i >= self.exp_to_reach.len() {
            return None;
        }
        Some(self.exp_to_reach[i])
    }

    pub fn cumulative_xp(&self, level: i32, exp_in_level: i64) -> Option<i64> {
        if level < 1 || (level as usize) >= self.cum_exp.len() {
            return None;
        }
        Some(self.cum_exp[level as usize] + exp_in_level)
    }

    pub fn summary(&self) -> String {
        format!("levels 2..{}", self.max_level)
    }
}

pub fn parse(vars: &HashMap<String, String>, fetched_at: DateTime<Utc>) -> Result<Catalog, String> {
    let mut exp = HashMap::new();
    let mut max_level = 0i32;
    for (key, raw) in vars {
        let Some(rest) = key.strip_prefix("exp_lvl") else {
            continue;
        };
        if rest.is_empty() || !rest.bytes().all(|b| b.is_ascii_digit()) {
            continue;
        }
        let level: i32 = rest
            .parse()
            .map_err(|_| format!("catalog: {key} is not a level"))?;
        if !(1..=LEVEL_CAP).contains(&level) {
            continue;
        }
        let v: i64 = raw
            .parse()
            .map_err(|_| format!("catalog: {key} = {raw:?} is not a number"))?;
        if v < 0 {
            return Err(format!("catalog: {key} = {v} is negative"));
        }
        if level > max_level {
            max_level = level;
        }
        exp.insert(level, v);
    }
    if max_level < 2 {
        return Err("catalog: the feed carries no exp_lvl table".into());
    }
    let mut exp_to_reach = vec![0i64; (max_level as usize) + 1];
    for level in 2..=max_level {
        let Some(v) = exp.get(&level) else {
            return Err(format!(
                "catalog: exp_lvl{level} is missing, so the XP table has a hole \
                 and cumulative totals past level {} would be wrong",
                level - 1
            ));
        };
        exp_to_reach[level as usize] = *v;
    }
    let mut c = Catalog {
        schema_version: SCHEMA_VERSION,
        fetched_at,
        max_level,
        exp_to_reach,
        cum_exp: Vec::new(),
    };
    c.finish();
    Ok(c)
}

fn fetch(
    url: &str,
    user_agent: &str,
    timeout: Duration,
    fetched_at: DateTime<Utc>,
) -> Result<Catalog, String> {
    let agent = ureq::AgentBuilder::new().timeout(timeout).build();
    let resp = agent
        .get(url)
        .set("User-Agent", user_agent)
        .call()
        .map_err(|e| format!("catalog: {e}"))?;
    if resp.status() != 200 {
        return Err(format!("catalog: HTTP {}", resp.status()));
    }
    let mut raw = Vec::new();
    resp.into_reader()
        .take(MAX_BODY)
        .read_to_end(&mut raw)
        .map_err(|e| format!("catalog: reading body: {e}"))?;
    let body = String::from_utf8_lossy(&raw);
    if dfclient::looks_like_html(&body) {
        return Err(
            "catalog: got an HTML page instead of the feed (Cloudflare or an outage)".into(),
        );
    }
    if raw.len() < MIN_BODY {
        return Err(format!(
            "catalog: body is only {} bytes, far short of the ~1MB feed; \
             treating it as truncated rather than parsing a partial table",
            raw.len()
        ));
    }
    let vars = dfclient::parse_flash(&body).map_err(|e| format!("catalog: {e}"))?;
    parse(&vars, fetched_at)
}

fn load_file(path: &Path) -> Result<Option<Catalog>, String> {
    let data = match std::fs::read(path) {
        Ok(d) => d,
        Err(e) if e.kind() == std::io::ErrorKind::NotFound => return Ok(None),
        Err(e) => return Err(e.to_string()),
    };
    match serde_json::from_slice::<Catalog>(&data) {
        Ok(mut c) if c.schema_version == SCHEMA_VERSION && c.exp_to_reach.len() >= 3 => {
            c.finish();
            Ok(Some(c))
        }
        _ => {
            let aside = format!("{}.corrupt-{}", path.display(), Utc::now().timestamp());
            std::fs::rename(path, &aside)
                .map_err(|e| format!("catalog cache unusable and could not be moved aside: {e}"))?;
            Ok(None)
        }
    }
}

fn save_file(path: &Path, c: &Catalog) -> Result<(), String> {
    if let Some(dir) = path.parent() {
        std::fs::create_dir_all(dir).map_err(|e| e.to_string())?;
    }
    let data = serde_json::to_vec(c).map_err(|e| e.to_string())?;
    let tmp = path.with_extension("json.tmp");
    std::fs::write(&tmp, data).map_err(|e| e.to_string())?;
    std::fs::rename(&tmp, path).map_err(|e| e.to_string())
}

pub fn ensure(
    path: &Path,
    feed_url: &str,
    user_agent: &str,
    max_age: Duration,
    timeout: Duration,
    now: DateTime<Utc>,
) -> Result<Catalog, String> {
    let cached = load_file(path)?;
    if let Some(c) = cached.as_ref() {
        if now >= c.fetched_at {
            let age = (now - c.fetched_at).to_std().unwrap_or(Duration::ZERO);
            if age < max_age {
                return Ok(c.clone());
            }
        }
    }
    match fetch(feed_url, user_agent, timeout, now) {
        Ok(fresh) => {
            let _ = save_file(path, &fresh);
            Ok(fresh)
        }
        Err(err) => {
            if let Some(c) = cached {
                Ok(c)
            } else {
                Err(err)
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::dfclient::parse_flash;

    fn load_fixture() -> Catalog {
        let raw = std::fs::read_to_string(
            Path::new(env!("CARGO_MANIFEST_DIR")).join("testdata/allstats.txt"),
        )
        .unwrap();
        let vars = parse_flash(&raw).unwrap();
        parse(&vars, Utc::now()).unwrap()
    }

    #[test]
    fn parses_the_real_feed() {
        let c = load_fixture();
        assert_eq!(c.max_level, 415);
        assert_eq!(c.exp_to_reach[2], 125);
        assert_eq!(c.exp_to_reach[415], 176_000_000);
        assert_eq!(c.exp_to_reach.len(), 416);
    }

    #[test]
    fn cumulative_xp_is_continuous_across_level_up() {
        let c = load_fixture();
        for level in [1, 2, 5, 50, 200, 414] {
            let needed = c.exp_needed(level).unwrap();
            let before = c.cumulative_xp(level, needed - 1).unwrap();
            let after = c.cumulative_xp(level + 1, 0).unwrap();
            assert_eq!(after - before, 1, "level {level}");
        }
        let mut prev = -1i64;
        for level in 1..=c.max_level {
            let got = c.cumulative_xp(level, 0).unwrap();
            assert!(got > prev, "level {level}");
            prev = got;
        }
    }

    #[test]
    fn exp_needed_at_the_cap() {
        let c = load_fixture();
        assert!(c.exp_needed(415).is_none());
        assert_eq!(c.exp_needed(414).unwrap(), 176_000_000);
    }

    #[test]
    fn hole_in_the_table_is_an_error() {
        let mut vars = HashMap::new();
        vars.insert("exp_lvl2".into(), "10".into());
        vars.insert("exp_lvl4".into(), "40".into());
        assert!(parse(&vars, Utc::now()).unwrap_err().contains("exp_lvl3"));
    }
}
